### ISSUE TITLE
unable to reproduce imagenet linear classification

### ISSUE THREAD

**Issue by PrateekMunjal** [orashi/Fast-MoCo#2]
Hi, thanks for your amazing work and excellently written codebase. 

TL;DR: We repeated our runs two times and we achieved an accuracy of 73.08% (attached logfile - baseline_log.txt) and 73.11% when the linear head is trained with Imagenet. However, we observe that in the main paper the performance is reported at 73.5% which is 0.4% higher and is **significant** on the scale of the Imagenet valset. 

<img width="611" alt="paper_results" src="https://user-images.githubusercontent.com/7415240/189839788-58f94397-31ae-4dba-9032-f326d46c6a5f.png">

At the same time, the self-supervised model checkpoint shared on github [here (https://drive.google.com/file/d/12ZEKiUg8ep2LgX5cJbEELmPHJRcFEMrR/view)  does reproduce the said result. 

So we suspect that there can be three possibilities leading to irreproducible results: 
     1) either there is an issue in the way we train our self-supervised model 
or 2) there is an issue in training the linear head. 
or 3) Was there a variance of 0.5% in performance if we simple repeat self-supervised training?

The second issue in training the linear classifier looks unlikely the reason because with the exact same settings we were able to achieve 73.51% if we start from the checkpoint provided on GitHub (corresponding to 100 epochs). We also provide the log file for reference- baseline_github_log.txt

Another reason, we can think is that it's possible that there is a difference in the way we train our self-supervised model. We share our training logs for your reference - log_self_sup.txt

Kindly let us know by comparing our logs to your self-sup training logs (which you used for reporting the results on the paper) if there is any major change.

Looking forward to a positive reply asap, and, once again thanks for your amazing work :)

[baseline_github_log.txt](https://github.com/orashi/Fast-MoCo/files/9554842/baseline_github_log.txt)
[log_self_sup.txt](https://github.com/orashi/Fast-MoCo/files/9554837/log_self_sup.txt)
[baseline_log.txt](https://github.com/orashi/Fast-MoCo/files/9554846/baseline_log.txt)

**Comment by orashi** [comment]
Hi Prateek,

Thanks for your interest to our work!

I believe somehow a difference between this repo and your actual training code/configuration is the cause of the performance gap. A 0.4 difference in Top-1 is beyond the reasonable variance range based on our past experiments (without saying you have hit 73.1 twice). 

I have compared our logs (one corresponds to the reported result and one we rerun based on the code as in this repo) with the one provided by you:
![download](https://user-images.githubusercontent.com/13995456/190872885-7281ed19-83c7-4fa5-b971-e7aa918ec995.png)
The curves are close but still we can see a difference between yours and two runs by us. 

Based on the configuration printed within your log. I spotted only one difference that could potentially cause the difference. Specifically, you are using 'distributed_iteration' sampler instead of our default 'distributed_epoch' sampler. If your code is the same as ours in terms of sampler implementation, 'distributed_iteration' will allow an image to appear multiple/zero times within an epoch as it shuffles index over epoch*dataset_len list instead of repeatedly shuffling dataset_len lists. We never tested our method on this sampler, and I haven't seen any SSL method that uses this kind of sampler in their experiments, so the impact of this is quite unknown. 

May you try to switch to 'distributed_epoch' and run again if this is the case?

If it doesn't work out, please provide your code or key differences in your implementation so I can help you.

codes and logs for the plot:
[issue.zip](https://github.com/orashi/Fast-MoCo/files/9592448/issue.zip)

**Comment by orashi** [comment]
BTW, you might notice some naming differences and configurations not present in this repo for the logs I attached, It is just caused by some code cleaning I did for the public repo and does not affect the actual execution.

**Comment by PrateekMunjal** [comment]
Thanks for your detailed feedback :) I will surely get back on this.

**Comment by PrateekMunjal** [comment]
Hi @orashi, thanks for your suggestion to use the epoch based sampler instead of using iteration based sampler. We found the results reproduce using epoch based sampler, precisely, we got a top1 accuracy of 73.55.

![acc1_val](https://user-images.githubusercontent.com/7415240/191822277-fc7e1eb3-633d-407e-9b53-613e04ed08a4.svg)
